src/librustc_mir/borrow_check/mod.rs

```diff
@@ -1524,47 +1598,89 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
+
+                        let item_msg = match self.describe_place(place) {
+                            Some(name) => format!("immutable item `{}`", name),
+                            None => "immutable item".to_owned(),
+                        };
+
+                        let mut err = self.tcx
+                            .cannot_borrow_path_as_mutable(span, &item_msg, Origin::Mir);
+                        err.span_label(span, "cannot borrow as mutable");
+
+                        if place != place_err {
+                            if let Some(name) = self.describe_place(place_err) {
+                                err.note(
+                                    &format!("Value not mutable causing this error: `{}`", name)
+                                );
+                            }
+                        }
 
-                if place != place_err {
-                    if let Some(name) = self.describe_place(place_err) {
-                        err.note(&format!("Value not mutable causing this error: `{}`", name));
+                        err.emit();
                     }
                 }
-
-                err.emit();
-            },
+            }
             Reservation(WriteKind::Mutate) | Write(WriteKind::Mutate) => {
-                if let Err(place_err) = self.is_mutable(place, is_local_mutation_allowed) {
-                    error_reported = true;
-
-                    let item_msg = match self.describe_place(place) {
-                        Some(name) => format!("immutable item `{}`", name),
-                        None => "immutable item".to_owned(),
-                    };
-
-                    let mut err = self.tcx.cannot_assign(span, &item_msg, Origin::Mir);
-                    err.span_label(span, "cannot mutate");
-
-                    if place != place_err {
-                        if let Some(name) = self.describe_place(place_err) {
-                            err.note(&format!("Value not mutable causing this error: `{}`", name));
+                match self.is_mutable(place, is_local_mutation_allowed) {
+                    Ok((Place::Local(local), mut_allowed)) => {
+                        if mut_allowed != LocalMutationIsAllowed::Yes {
+                            // If the local may be initialized, and it is now currently being
+                            // mutated, then it is justified to be annotated with the `mut`
+                            // keyword, since the mutation may be a possible reassignment.
+                            let mpi = self.move_data.rev_lookup.find_local(*local);
+                            if flow_state.inits.contains(&mpi) {
+                                self.used_mut.insert(*local);
+                            }
+                        }
+                    }
+                    Ok((Place::Projection(ref proj), _mut_allowed)) => {
+                        if let Some(field) = self.is_upvar_field_projection(&proj.base) {
```

Comment:
So, the code doesn't work because of this line -- it's checking whether the _base projection_ is an upvar field projection, which clearly isn't the case, so instead we need to check the projection itself. I don't know if this would change anything else, but I'm going to try making this small change to see what results.